Refactoring models representation

Halbaroth opened the pull request:
This PR changes the way Alt-Ergo stores models. 

Before the PR, Alt-Ergo didn't produce a printable model in memory but performed some basic computations in the model printers. 

Now, Alt-Ergo will produce in memory a printable model. The PR mostly consists in moving some codes from `Models.ml` to `ModelMap.ml` or `Uf.ml`.

- All the require computation is now done in the `output_concrete_model` of the union-find module (included the computation for records).
- All the printings stuff are in `ModelMap.ml`.
- We use `Expr.t` to store model values.
- The only reason I kept `Models.ml` is because we are the printer of `constraint model` here. This will be remove later.

Besides, as I notice, the symbols we define in models are all `Sy.Name`. Thus I create an new module to manage identifiers `Id.ml` and I store identifiers instead of symbols in models.

Halbaroth committed:
Refactoring model

Halbaroth committed:
Creating a module for identifiers

As we need to use identifiers in both `Symbols` and `Value` of models, I
create a new module `Id.ml` which contains namespaces for internal
identifiers and abstract values.

Halbaroth committed:
Clean-up printers

Halbaroth committed:
Reset the cache at the start of generation model

The call to `Cache.reset` wasn't in the appropriate function and
reset the cache after computing each model value.

Halbaroth committed:
Produce an abstract value for abstract type.

We cannot produce a concrete value for a symbol whose the type is an
abstract value. For instance, consider the input file:

```
(set-logic ALL)
(set-option :produce-models true)
(declare-sort s 0)
(declare-const a s)
(declare-const b s)
(assert (= a b))
(check-sat)
(get-model)
```

Alt-Ergo outputs `unknown`, which makes sense as we don't know anything
about the abstract type `s`. As `a` and `b` are user-defined, we have to
print something in the model about them but we cannot! Thus, we have to
print abstract values for them.

Halbaroth committed:
Restoring record model generation

Halbaroth committed:
Print the same abstract value of equal semantic values.

If we have the equation `a = b` for two terms `a` and `b` and we
produce an abstract value for `a`, we have to use the same abstract
value for `b` in the model.

Halbaroth committed:
Prefix unused arguments of model values

Prefix unused arguments of model values by an underscore. In the current
codebase, the only situation in which such arguments can appear is while
printing constants.

Halbaroth committed:
Restoring objectives in interpretation

Halbaroth committed:
Restoring assertion mode for models and clean-up

Halbaroth committed:
Use `Seq.is_empty` from `stdcompat`

Halbaroth committed:
Cleanup and a bit of documentation

Halbaroth committed:
Promote tests

Halbaroth committed:
Linter

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1020,7 +1026,9 @@ let model_repr_of_term t env mrepr unbounded =
     match unbounded with
-    | Some string_repr -> (rep, string_repr), mrepr
+    | Some string_repr ->
+      assert (is_a_good_model_value rep && is_const_term rep);
```
Can you add a quick comment explaining the reason for this assertion?

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1030,69 +1038,157 @@ let model_repr_of_term t env mrepr unbounded =
-      let e = X.choose_adequate_model t rep cls in
-      e, ME.add t e mrepr
+      let rep, string_repr = X.choose_adequate_model t rep cls in
+      assert (is_a_good_model_value rep && is_const_term rep);
```
Same here

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1030,69 +1038,157 @@ let model_repr_of_term t env mrepr unbounded =
-    env t ((fprofs, cprofs, carrays, mrepr) as acc) unbounded =
+let compute_concrete_model_of_val env t ((mdl, mrepr) as acc) unbounded =
   let { E.f; xs; ty; _ } = E.term_view t in
   (* Keep record constructors because models.ml expects them to be there *)
```
Isn't this comment linked to the `is_forbidden_symbol` ?

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1030,69 +1038,157 @@ let model_repr_of_term t env mrepr unbounded =
+      let arg_vals, arg_tys, mrepr =
+        List.fold_left
+          (fun (arg_vals, arg_tys, mrepr) arg ->
+             let rep_arg, mrepr = model_repr_of_term arg env mrepr unbounded in
```
There was an assertion here, why is it removed?

Halbaroth replied:
I moved it in `model_repr_of_term` because we need to enforce this invariant twice.

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1030,69 +1038,157 @@ let model_repr_of_term t env mrepr unbounded =
+          )
+          ([], [], mrepr) (List.rev xs)
+      in
+      let ret_rep, mrepr = model_repr_of_term t env mrepr unbounded in
```
Same here

Halbaroth replied:
Same answer :D

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1030,69 +1038,157 @@ let model_repr_of_term t env mrepr unbounded =
 
-let save_cache () =
-  LX.save_cache ()
+module Cache = struct
```
Can you add a short documentation for this module?

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1107,9 +1203,38 @@ module MED = Map.Make
+(* The model generation is known to be imcomplete for FPA and Bitvector
+   theories. *)
+let is_suspicious_symbol = function
+  | Symbols.Op (Float | Fixed | Abs_int | Abs_real | Sqrt_real
```
Symbols -> Sy

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1107,9 +1203,38 @@ module MED = Map.Make
+         (* We don't store names defined by the user. *)
+         acc
+       | _ ->
+         let suspicious = if is_suspicious_symbol f then true else suspicious in
```
```suggestion
         let suspicious = is_suspicious_symbol f || suspicious
```

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1122,29 +1247,72 @@ let compute_concrete_model ?(optimized_splits=Util.MI.empty) env =
          | Unknown -> acc
       ) optimized_splits (SetX.empty, SetX.empty, SetX.empty)
   in
+
```
```suggestion
```

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1122,29 +1247,72 @@ let compute_concrete_model ?(optimized_splits=Util.MI.empty) env =
-         acc
-    ) (terms env)
-    (ModelMap.empty, ModelMap.empty, ModelMap.empty, ME.empty)
+
```
```suggestion
```

Stevendeo commented on src/lib/reasoners/uf.ml:
```diff
@@ -1122,29 +1247,72 @@ let compute_concrete_model ?(optimized_splits=Util.MI.empty) env =
+          let Expr.{ f; _ } = Expr.term_view t in
+          match f with
+          | Sy.Name (id, _, _) -> id
+          | _ -> assert false
```
Quick explanation about this assertion

Stevendeo commented on src/lib/structures/modelMap.ml:
```diff
@@ -30,100 +30,179 @@
-
-let is_empty { values; _ } = P.is_empty values
+let add ((_, arg_tys, _) as sig_) arg_vals ret_val { values; suspicious } =
+  assert (List.compare_lengths arg_tys arg_vals = 0);
```
Add some documentation about this assertion

Halbaroth replied:
I replaced it by an `invalid_argument` as explained offline ;D

Stevendeo commented on src/lib/util/util.mli:
```diff
@@ -119,3 +119,5 @@ val print_list_pp:
 
 val internal_error : ('a, Format.formatter, unit, 'b) format4 -> 'a
+
+val quoted_string : string -> string
```
```suggestion
(** Returns the string in argument as a smtlib2 valid identifier (either itself or quoted). *)
val quoted_string : string -> string
```

Stevendeo commented on src/lib/structures/modelMap.ml:
```diff
@@ -30,100 +30,179 @@
+
+let pp ppf {values; suspicious} =
+  if suspicious then begin
+    Fmt.pf ppf "; This model is a best-effort. It includes symbols
+        for which model generation is known to be incomplete. @."
```
```suggestion
    Fmt.pf ppf "; This model is a best-effort. It includes symbols \
                for which model generation is known to be incomplete.@."
```

Halbaroth committed:
Update src/lib/structures/modelMap.ml

Co-authored-by: Stevendeo <de.oliveira.steven@gmail.com>

Halbaroth committed:
Update src/lib/util/util.mli

Co-authored-by: Stevendeo <de.oliveira.steven@gmail.com>

Halbaroth committed:
Update src/lib/reasoners/uf.ml

Co-authored-by: Stevendeo <de.oliveira.steven@gmail.com>

Halbaroth committed:
Update src/lib/reasoners/uf.ml

Co-authored-by: Stevendeo <de.oliveira.steven@gmail.com>

Halbaroth committed:
Review changes

Halbaroth committed:
Unused argument in the test

Halbaroth committed:
Promote tests

bclement-ocp commented on tests/models/uf/uf2.models.expected:
```diff
@@ -1,7 +1,7 @@
 unknown
 (
-  (define-fun f ((arg_0 Int)) Int (ite (= arg_0 0) 2 0))
+  (define-fun f ((arg_0 Int)) Int (ite (= arg_0 (- 2)) 0 2))
```
Why did this model change?

bclement-ocp commented on tests/float/test_float2.models.expected:
```diff
@@ -1,5 +1,7 @@
 
 unknown
+; This model is a best-effort. It includes symbols
+; for which model generation is known to be incomplete.
```
Note that after #886 model generation should only be incomplete when the FPA prelude is disabled — which I don't think is a configuration we want to support.

Halbaroth replied:
For this one, it seems the test is running without enabling the `fpa` theory:
```
  (rule
   (target test_float2.models_tableaux.output)
   (deps (:input test_float2.models.smt2))
   (package alt-ergo)
   (action
    (chdir %{workspace_root}
     (with-stdout-to %{target}
      (ignore-stderr
       (with-accepted-exit-codes 0
        (run %{bin:alt-ergo}
             --timelimit=2
             --enable-assertions
             --output=smtlib2
             --frontend dolmen
             --sat-solver Tableaux
             %{input})))))))
```

bclement-ocp replied:
My point is that we don't inspect the preludes for the "known to be incomplete" message and that I think we should never print it for the `float` function (rather than always print it), because we expect users to use them with the preludes (or to know what they are doing).

(BTW the fact that we reply `"unknown"` and not `"sat"` is also an indication that the model might be incomplete)

Halbaroth replied:
I don't understand your point. I don't know if I'm supposed to print something if the user tries to produce models in presence of FPA symbols without loading the FPA theory. I guess we should print something because users can forget flags.

bclement-ocp replied:
My point is that it is fine to not print anything since replying "unknown" is already saying that the model may be incomplete. At least we should remove this when #917 is taken care of.

bclement-ocp commented on tests/bitv/testfile-bv2nat-models.dolmen.expected:
```diff
@@ -1,5 +1,7 @@
 
 unknown
+; This model is a best-effort. It includes symbols
+; for which model generation is known to be incomplete.
```
After #886 model generation for bv2nat and int2bv should be complete.

bclement-ocp commented on src/lib/util/util.mli:
```diff
@@ -119,3 +119,7 @@ val print_list_pp:
+
+(** Returns the string in argument as a smtlib2 valid identifier
+    (either itself or quoted). *)
+val quoted_string : string -> string
```
Note for the future: this should be replaced by the appropriate functions in Dolmen once Gbury/dolmen#198 is merged.

Halbaroth replied:
I will add a comment in the code.

bclement-ocp commented on src/lib/structures/modelMap.mli:
```diff
@@ -28,33 +28,52 @@
-    of expressions matching the P.key type (V).
-*)
+type sy = Id.t * Ty.t list * Ty.t
+(** Typed symbol of function. *)
```
I am not sure how to link the types in the tuple to the description "typed symbol of function". I believe this is: `(symbol, args_ty, ret_ty)` tuples where `symbol` is the function name, `args_ty` the type of arguments, `ret_ty` the return type?

Halbaroth replied:
I add a documentation to explain the meaning of its fields.

bclement-ocp commented on src/lib/structures/modelMap.mli:
```diff
@@ -28,33 +28,52 @@
-                        (Expr.t * (Shostak.Combine.r * string)) list *
-                        (Shostak.Combine.r * string)
+    | `Constant of Hstring.t
+    (** A string representation of a semantic value. *)
```
Why do we use a string representation here? I thought the point of the PR was to build a structured representation of the model prior to printing, but if we just have strings everywhere I don't really understand the difference.

Besides, I am not sure I understand why we need these strings to be interned (i.e. why is this `Hstring.t` rather than simply `string`), can you clarify?

Halbaroth replied:
I will remove the `Hstring.t`. As I explained, this is not a definitive version of the model representation.

bclement-ocp commented on src/lib/structures/modelMap.mli:
```diff
@@ -28,33 +28,52 @@
-module P : Map.S with type key =
-                        Symbols.t * Ty.t list * Ty.t
+module Value : sig
+  type simple = [
```
Why is this a polymorphic variant rather than a regular sum type?

Halbaroth replied:
Actually, I didn't want to use a different constructor for abstract of array and abstract of anything. I rolled back to a previous version and now it's just a classical ADT with two constructors.

bclement-ocp commented on src/lib/structures/modelMap.mli:
```diff
@@ -28,33 +28,52 @@
 
-type key = P.key
-type elt = V.t
+  type record = Hstring.t * simple list
```
Records can contain other records, why are the fields of type `simple`? I guess this is the same question as above wrt using strings for constants inside `simple`.

Halbaroth replied:
Fixed

bclement-ocp commented on src/lib/structures/modelMap.mli:
```diff
@@ -28,33 +28,52 @@
+    | `Abstract of sy
+    (** An unique abstract array value. *)
+
+    | `Store of array * Hstring.t * Hstring.t
```
Same as above; why are `key` and `value` strings rather than structured representations?

Halbaroth replied:
Fixed

bclement-ocp commented on src/lib/structures/modelMap.mli:
```diff
@@ -28,33 +28,52 @@
+    (** A string representation of a record definition. *)
 
-val iter : (key -> elt -> unit) -> t -> unit
+    | simple
```
This type allows both `` `Abstract sy`` and `` `Array (`Abstract sy)`` and I am not sure I understand the difference between these two.

Halbaroth replied:
It was done on purpose. I rolled back to a previous version.

bclement-ocp commented on src/lib/structures/modelMap.mli:
```diff
@@ -28,33 +28,52 @@
 
-val is_empty : t -> bool
+val empty : suspicious:bool -> t
+(** An empty model. *)
```
Should briefly explain the `suspicious` argument here.

bclement-ocp commented on src/lib/structures/modelMap.ml:
```diff
@@ -30,100 +30,182 @@
-          with
-          | Exit -> assert (!c <> 0); !c
-          | Invalid_argument _ -> assert false
+type sy = Id.t * Ty.t list * Ty.t [@@deriving ord]
```
The order on these values has changed, not sure if that was intentional (also unsure whether the order matters or not here, but since it did not use lexicographic ordering initially it may).

Halbaroth replied:
I think it doesn't matter. This order is only use for the printing.

bclement-ocp commented on src/lib/structures/modelMap.ml:
```diff
@@ -30,100 +30,182 @@
+  let empty = M.empty
+  let add = M.add
+
+  module Fiber = struct
```
While the term "fiber" may be common on some areas of mathematics, it is not in the SMT community; if you want to use that term, please include a short comment explaining what it means.

bclement-ocp commented on src/lib/structures/modelMap.ml:
```diff
@@ -30,100 +30,182 @@
+          (and (= arg_1 x2)
+            ... (= arg_n x_n).
+    *)
+    let rec pp_args ctr ppf = function
+      | [] -> ()
+      | [arg] ->
+        Fmt.pf ppf "(= arg_%i %a)" ctr Value.pp arg
+      | arg :: args ->
+        Fmt.pf ppf "(and (= arg_%i %a) %a)"
+          ctr
+          Value.pp arg
+          (pp_args (ctr + 1)) args
```
Note that the `and` SMT-LIB constructor is n-ary, it would be more readable to generate `(and (= arg_0 x1) (= arg_1 x2) (= arg_2 x3))` rather than nesting `and`s.

Halbaroth replied:
I didn't know, I will change it :)

bclement-ocp commented on src/lib/structures/modelMap.ml:
```diff
@@ -30,100 +30,182 @@
+        | Cons (args, seq) ->
+          Fmt.pf ppf "(or %a %a)"
+            (pp_args 0) args
+            aux seq
```
The `or` constructor is also n-ary.

bclement-ocp commented on src/lib/structures/modelMap.ml:
```diff
@@ -30,100 +30,182 @@
+            (pp_args 0) args
+            aux seq
+      in
+      aux ppf (to_seq fiber)
```
I am not sure the overhead of converting to a sequence is needed here; we could just inspect the cardinal and iterate the set directly.

Halbaroth replied:
I didn't known `Fmt.iter_bindings` and `Fmt.iter` ;). I replaced them.

bclement-ocp commented on src/lib/structures/modelMap.ml:
```diff
@@ -30,100 +30,182 @@
-          with
-          | Exit -> !c
-          | Invalid_argument _ -> List.length l1 - List.length l2
+      type t = sy [@@deriving ord]
```
It feels icky to use `@@deriving ord` to provide the arguments to a functor, it feels very implicit. Defining `let compare = compare_sy` explicitly would be more explicit.

bclement-ocp commented on src/lib/structures/modelMap.ml:
```diff
@@ -30,100 +30,182 @@
+
+let empty ~suspicious = { values = P.empty; suspicious }
+
+let pp_named_arg_ty ~unused ppf (arg_name, arg_ty) =
+  let pp_unused ppf unused = if unused then Fmt.pf ppf "_" else () in
+  Fmt.pf ppf "(%aarg_%i %a)" pp_unused unused arg_name Ty.pp_smtlib arg_ty
```
If we want to add `_` to the unused arguments, we should either do it properly or not do it. Currently we only add `_` to the arguments if they are *all* unused, which is a middle ground that I don't think makes much sense; I would prefer we did nothing here.

Halbaroth replied:
Our expressions to print the graph have a very particular form. If we use one argument in our expression, we use all of them.

bclement-ocp replied:
Ah, that is correct indeed!

bclement-ocp commented on src/lib/structures/modelMap.ml:
```diff
@@ -30,100 +30,182 @@
+    Fmt.pf ppf "; This model is a best-effort. It includes symbols\n\
+                ; for which model generation is known to be incomplete.@."
+  end;
+  Fmt.pf ppf "@[<v 2>(%a@]@,)" pp_seq (P.to_seq values)
```
We can check `P.is_empty` and then `Fmt.iter P.iter` rather than converting to a sequence.

bclement-ocp commented on src/lib/frontend/models.mli:
```diff
@@ -38,15 +38,11 @@ type objective_value =
+  model : ModelMap.t;
   objectives: (Expr.t * objective_value) Util.MI.t;
-  terms_values : (Shostak.Combine.r * string) Expr.Map.t
+  terms_values : (Shostak.Combine.r * Hstring.t) Expr.Map.t
```
Not sure why we need the values to be interned `Hstring.t`? Also, shouldn't we be using `Value.t` here?

Halbaroth replied:
It's not relevant anymore as we use `Expr.t` to store model values now.

bclement-ocp commented on src/lib/reasoners/uf.ml:
```diff
@@ -999,100 +999,206 @@ let assign_next env =
+end
+
+let is_forbidden_symbol f ty =
   (* Keep record constructors because models.ml expects them to be there *)
```
The corresponding code that expects the record constructors to be there in `models.ml` seems to have been removed, was it intentional?

Halbaroth replied:
Uhm I'm not sure. The function `is_forbidden_symbol` is used to filter the symbols on which we will iterate during the model generation. I think the comment means we have to keep `Access` symbol, even if this symbol is interpreted by the theory of records, otherwise we cannot construct the model value for this theory.

bclement-ocp replied:
I wrote this comment. It refers to the fact that in we (used to) need to preserve the record constructors for `check_records` et al in `models.ml`, but those functions have been removed. This means that either some stuff will no longer be working, or the new code in `uf.ml` is meant to replace those functions and we don't need to keep the record constructors here anymore (probably the latter).

Halbaroth replied:
This comment isn't relevant anymore as we don't need to fold on all the terms in the union-find to produce model values for record symbols.

bclement-ocp commented on src/lib/reasoners/uf.ml:
```diff
@@ -999,100 +999,206 @@ let assign_next env =
-  LX.save_cache ()
+(* Caches used during the computation of the model.
+   All these caches are flushed before generating a new model. *)
+module Cache = struct
```
Is there a reason why the caches are stored in global variables? It would be cleaner (and thread-safe) to have a type that bundles the hash tables and is passed to `compute_concrete_model_of_val`.

Halbaroth replied:
No reason, I will create the cache in `compute_concrete_model` directly.

Halbaroth replied:
I put caches in the closure of the function `extract_concrete_model`.

bclement-ocp commented on src/lib/reasoners/uf.ml:
```diff
@@ -999,100 +999,206 @@ let assign_next env =
+        end
+
+      | Sy.Op Sy.Set, _, _ ->
+        (* As arrays are immutable, the result of the set operator is
+           never a user-defined array and has to be ignored. *)
```
This is not true. The result of the set operator can be equal to a user-defined array, as in #929 (in fact this may or may not be the source of #929)

Halbaroth replied:
My comment was misleading. In fact `Set` symbols occur here by accident because `X.is_my_symbol` isn't the appropriate function to filter terms here. I just removed the comment and I'll do another PR in which a create a better predicate for this use case.

bclement-ocp commented on src/lib/reasoners/uf.ml:
```diff
@@ -999,100 +999,206 @@ let assign_next env =
-      else
-        ModelMap.add (f, tys, ty) (xs, rep) fprofs, cprofs, carrays,
-        mrepr
+      | Sy.Name (id, _, _), [], Ty.Trecord trecord ->
```
Are we sure nothing special needs to happen for functions returning records?

Halbaroth replied:
I will check but as it checks there is no argument and a function without argument is the same as a constant in the SMT-LIB standard, it should be ok.

bclement-ocp replied:
What I mean is that we are calling `store_record_sig` for constant records, but not for functions that return records, and I am wondering if we should do something similar for functions that return records.

See e.g. #412 where there is an uninterpreted function that returns a record and where Alt-Ergo gets confused. On a side note, I thought I had fixed the first half of the bug (treating all functions with a single record argument as constructors) but apparently I never made that PR, oops :innocent: . Would be worth checking what happens there with this implementation.

Halbaroth replied:
I fixed it and add some tests. This PR solves #412 .

bclement-ocp commented on src/lib/reasoners/uf.ml:
```diff
@@ -1107,9 +1213,38 @@ module MED = Map.Make
+(* The model generation is known to be imcomplete for FPA and Bitvector
+   theories. *)
+let is_suspicious_symbol = function
+  | Sy.Op (Float | Fixed | Abs_int | Abs_real | Sqrt_real
+          | Sqrt_real_default | Sqrt_real_excess
+          | Real_of_int | Int_floor | Int_ceil
+          | Max_int | Max_real | Min_int | Min_real
+          | Pow | Integer_log2 | Int2BV _ | BV2Nat
+          | BVand | BVor | Integer_round) -> true
```
After #886, `BV2Nat`, `Int2BV` and `Pow` should no longer be treated as suspicious. Further, `Float`, `Abs_int`, `Abs_real`, `Sqrt_real_default`, `Sqrt_real_excess`, `Real_of_int`, `Int_floor`, `Int_ceil`, `Max_int`, `Max_real`, `Min_int`, `Min_real`, `Integer_log2` and `Integer_round` are now only suspicious if the FPA prelude is not loaded.

This is because we now propagate their arguments during model generation, which should prevent us from generating models with invalid values for these functions.

bclement-ocp reviewed (changes requested):
This PR feels like it stops halfway through: from reading the PR description I thought that we would be building a structured representation of the model in memory prior to printing, but we actually build a fairly shallow map where most things are strings. I think we should go all the way and keep the structured representation throughout, rather than build (and intern!) intermediate strings that are only used for printing; although there are maybe blockers to that I don't immediately see — thoughts?

> Besides, as I notice, the symbols we define in models are all Sy.Name. Thus I create an new module to manage identifiers Id.ml and I store identifiers instead of symbols in models.

I am not entirely convinced this is a good idea — I am not opposed to it but we may want to include things that are not `Sy.Name` in the models at some point (namely, values for partially interpreted functions such as division by zero) where we may regret this change.

(Unrelated thought: at some point we should probably use the `Id` module from Dolmen to represent identifiers as well)

Halbaroth committed:
Models for functions returning records

The previous implementation of `compute_concrete_model` doesn't support
properly model generation for functions returning records. This is a
quick fix.

Halbaroth committed:
Improve the type of `ModelMap.Value`

The previous type was not sufficient to represent all the possible model
values. In particular, we couldn't represent embedded records in other
record and we cannot use complex types for indices and values of functional
arrays.

Halbaroth commented:
> This PR feels like it stops halfway through: from reading the PR description I thought that we would be building a structured representation of the model in memory prior to printing, but we actually build a fairly shallow map where most things are strings. I think we should go all the way and keep the structured representation throughout, rather than build (and intern!) intermediate strings that are only used for printing; although there are maybe blockers to that I don't immediately see — thoughts?

As I explained in the description of this PR, the purpose of it is to solve the issue #833. Actually, this PR comes from a first commit I did in a old PR I've never pushed because we had to merge OptimAE first. I keep it as simple as possible because I want to merge it before #921.

In a next PR, I plan to produce models with the following ADT:
```
  type simple =
    | Int of Z.t
    | Real of Q.t
    | Bool of bool
    | Constructor of string

  type array =
    | AbstractArray of Id.t * Ty.t * Ty.t
    | Store of array * t * t

  and t =
    | Abstract of sy
    | Constant of simple
    | Array of array
    | Record of string * t list
```
This should be sufficient to perform evaluation of expression on the model, and so to perform checking  models.

Actually, I'm not really happy of the next commits of my old PR. The situation is complicated because we have a conflict between two designs: 

- First, the `semantic value` type is abstract, which means we cannot using directly the semantic values to get the model values because we have no access to their recursive structures in the `uf` module.
The case of records is particularly eloquent here. Generating the model value in `shostak` would be so easy. You just need to write a recursive function. This is what I did in my old PR. It works very fine with all the current theories but `arrays`...

- Second, arrays has no semantic value. In other words, we cannot retrieve directly the model value from the semantic values by a simple recursive call in `shostak`. Currently, we explore all the union-find in `uf` and retrieve the mode value by catching all the `get` expressions and accumulating values in caches.

There is  another annoying issue. The current function `choose_adequate_model` needs the class of the value for which we want a model value. For recursive values, as records or ADT, we will need more as we will perform recursive call. So the new `choose_adequate_model` requires the state of the union-find.... which means the best place to perform this computation is `uf` as it is actually? I don't know, may be we should create semantic values for arrays?

I fixed the issue you noticed with records and add some tests.

Halbaroth commented:
> I am not entirely convinced this is a good idea — I am not opposed to it but we may want to include things that are not `Sy.Name` in the models at some point (namely, values for partially interpreted functions such as division by zero) where we may regret this change.

If we need some other symbols in the futur, we can create a new type for it in `ModelMap`. A lot of the symbols of `Symbols` are fully interpreted and shouldn't appear in the signature of model values.

bclement-ocp commented on src/lib/structures/modelMap.mli:
```diff
@@ -28,33 +28,37 @@
-                        Symbols.t * Ty.t list * Ty.t
+module Value : sig
+  type array =
+    | AbstractArray of Id.t * Ty.t * Ty.t
```
You changed the name from `Abstract` to `AbstractArray` but my question remains: what is the difference between `Abstract (n, tys, ty)` and `Array (AbstractArray (n, tys, ty))`? I feel like just having `Store of t * t * t` should be enough here and would avoid this possible confusion.

Halbaroth replied:
Not relevant anymore.

bclement-ocp reviewed (commented):
I don't understand why these changes are needed to fix #833. #833 is about when we decide to print the model or not. This PR doesn't change that, it changes the way we store the model prior to printing it. Besides, the offending call to `Options.get_interpretation` that was the source of #833 is no longer present in `output_concrete_model`, so to me it looks like #833 is already fixed.

Anyways, arrays cannot have (meaningful) semantic values because the theory of arrays cannot have a Shostak solver, so I think we are stuck with a few contorsions for arrays specifically.

That doesn't mean we have to do the same contorsions for records. In fact, for Shostak theories we can (and probably should) use semantic values, as you suggest here and we also discussed a couple months ago. I don't think that the type of semantic values being abstract is a big issue; we just need to figure out the proper API to expose from the `Shostak` module.

I think that the best way might be to expose:
 - `children : r -> r list` that returns the list of direct alien children of a semantic value
 - `pp_model : r Fmt.t -> r Fmt.t` that prints a semantic value, using the first argument for its children

This should be enough for the union-find part to do its job. In fact, since model generation is done through the case split mechanism (`case_split ~for_model:true` is first called before extracting the model), we may only need `pp_model`?

This way, we can check with `X.is_a_leaf` -- if it returns true, this is an uninterpreted term (or AC symbol) and we can either print the associated abstract constant or (if it's an array) the structure reconstructed from UF. And we could get rid of most of the strings laying around…

(As an aside, I am not sure I understand the issue with `choose_adequate_model` and recursion. At the point we reach `choose_adequate_model` all values should be (conceptually) ground and have been assigned by the previous calls to `assign_value`. In fact, I think that arrays (and records, in the current implementation) are the only ones that truly need the equivalence class in `choose_adequate_model` — in my understanding any other theory needing this equivalence class is a symptom of a bug or a design issue.)

Halbaroth commented:
So I close this PR?

bclement-ocp commented:
I think this PR does some things right (even fixes some bugs!), even though it doesn't seem needed to fix #833. In particular, having a structured representation of models is useful (notably because for some theories such as arrays the solver doesn't really store the information in the appropriate way in the first place we we noted), even if we may want to change the way that representation is built in the future.

That said, if it is the first in a sequence of PRs that you are no longer happy with, maybe we should first take some time to brainstorm the API we want to have (my gut feeling is that it should be centered around *canonical semantic values*, not symbols, associated with an auxiliary mapping for uninterpreted terms that don't have ground semantic values, i.e. arrays, functions) to make sure we are going towards the direction we want and one that fixes the design bugs we can encounter with the current implementation.

Halbaroth committed:
Change review

Halbaroth committed:
rebase artefact

Halbaroth committed:
Use `Expr.t` to store model values

Use `Expr.t` values instead of strings to store the value of the
first-order model.

Add a new function `to_term_const` in the signature of `Shostak`.
Basically this function is the inverse function of `X.make` on the
constant terms only. The function always returns a term `t` such that
`Expr.is_const_term t` is `true`.

Notice that we need this function during model generation. Indeed, even
if the class of a semantic value in UF contains terms whose the `make`
is constant according to `X.is_constant`, these terms aren't necessary
constant according to `Expr.is_const_term`. For instance, the term `0 +
1` will become the semantic value `1` and we expect that
`X.to_term_const` returns `Some 1`.

Modify the definition of `Expr.is_const_term` (formerly named
`const_term`). The previous definition considered that the application
of constructor of an ADT to constant terms isn't a constant term. The
same went for record definitions. Now, there are constant too.
We have to check that this modification is correct for the

Notice that we keep the old definion `const_term` in the module `Expr`
but we don't expose it anymore. Indeed, this function is used to detect
constants in the smart constructor of the let bindings because the
definition of the depth of formulae have been tweaked to prevent
regressions. Modifying this function could be dangereous.

Halbaroth committed:
Poetry

Halbaroth committed:
Hide the caches in the closure of `compute_concrete_model`

Halbaroth committed:
Fix `to_const_term` in BV theory

Halbaroth committed:
Documentation in UF and fix BV symbols

Currently, we produce casesplits for bitvectors in order to generate
models but we cannot reasoning with `BVand`, `BVxor` and `BVor`. It
means we cannot add these symbols as interpreted symbols of the theory
BV (see the function `Bitv.is_mine_symbol`) but these symbols can appear
in the union-find environment. We have to ignore them while computing
the model in UF.

Halbaroth committed:
Documentation of the type sy

Halbaroth committed:
documentation

Halbaroth committed:
restoring the constraint printer

Halbaroth commented:
I update this PR to use `Expr.t` instead of strings in order to store model values. I also add a new function `to_const_term` in `sig.mli`. This function returns a constant interpreted term from a constant semantic value. Please refer to the commit message of [caae55f](https://github.com/OCamlPro/alt-ergo/pull/925/commits/caae55f0c7a40c6610dca07386d1a7611975e4e5) for more details.

Halbaroth committed:
Incorrect `is_internal` predicate

We have to consider every symbol names starting with a dot or a `at` as
an internal symbols. For instance the CDCL SAT solver produces internal
names prefixed by `.PROXY__`.

Halbaroth committed:
Simpler `to_const_term` in BV theory

We don't need to consider the case where a constant semantic value would
be a concatenation of constant simple terms. Now the canonizer eliminate
this case completely.

bclement-ocp commented on src/lib/frontend/models.mli:
```diff
@@ -32,16 +32,12 @@
-  arrays : ModelMap.t;
-  terms_values : (Shostak.Combine.r * string) Expr.Map.t
+  model : ModelMap.t;
+  terms_values : Expr.t Expr.Map.t
```
Nit: this should be `term_values`.

bclement-ocp commented on src/lib/frontend/models.ml:
```diff
@@ -245,221 +231,19 @@ module Pp_smtlib_term = struct
-  let fresh_counter = ref 0
-
-  let reset_counter () = fresh_counter := 0
+(* of module SmtlibCounterExample *)
```
Not sure what this means?

Halbaroth replied:
I guess it's a rebase artefact. It was used to delimit a module that doesn't exist anymore.

bclement-ocp commented on src/lib/reasoners/shostak.ml:
```diff
@@ -369,7 +386,14 @@ struct
     | Ite t -> ITE.is_constant t
-    | Ac _ | Term _ -> false
+    | Term t ->
+      begin
+        let Expr.{ f; xs; _ } = Expr.term_view t in
+        match f, xs with
+        | Symbols.(True | False | Void), [] -> true
+        | _ -> false
+      end
```
There should be a comment here explaining why this is not `Expr.is_const_term` (because other constants are semantic values).

Halbaroth replied:
In fact, we could use `Expr.is_const_term`. I didn't use it because when I wrote this code, the predicate `Expr.is_const_term` should accept variables as we use it in the smart constructor of let bindings. I decided to use two predicates and we don't need to accept variables with `Expr.is_const_term` anymore. Still, it makes sense to write a specific code here. I add a comment to explain the situation.

Halbaroth replied:
After simplifying the code `model_repr_of_term` in the union-find, we don't need to change `X.is_constant` anymore. But I think it would be better to have this invariant:
```ocaml
let mk, _ = X.make e in
if Expr.is_const_term e then assert (X.is_constant mk)
```
In other words, `X.make` sends model values to constant semantic values.

```ocaml 
  let is_constant r =
    match r.v with
    | Arith t -> ARITH.is_constant t
    | Records t -> RECORDS.is_constant t
    | Bitv t -> BITV.is_constant t
    | Arrays t -> ARRAYS.is_constant t
    | Enum t -> ENUM.is_constant t
    | Adt t -> ADT.is_constant t
    | Ite t -> ITE.is_constant t
    | Term t ->
      begin
        let Expr.{ f; xs; _ } = Expr.term_view t in
        match f, xs with
        | Symbols.(True | False | Void | Name _), [] -> true
        | _ -> false
      end
    | Ac _ -> false
```

bclement-ocp commented on src/lib/reasoners/sig.mli:
```diff
@@ -128,11 +128,7 @@ module type SHOSTAK = sig
-  *)
-  val choose_adequate_model : Expr.t -> r -> (Expr.t * r) list -> r * string
-
+  val to_const_term : r -> Expr.t option
```
Should be documented here.

Halbaroth replied:
I put the documentation in the second signature. I do a copy.

bclement-ocp commented on src/lib/reasoners/uf.ml:
```diff
@@ -999,95 +998,42 @@ let assign_next env =
+    let tt =
+      match cls with
+      | [] ->
+        (* We do this because terms of some semantic values created by CS
+           aren't created and added to the union-find. *)
+        X.to_const_term rep
```
Not sure what this comment means; terms of *most* semantic values created by Alt-Ergo are not added to `env.make`. I'd have expected `X.to_const_term rep` to always be the correct answer (except for arrays). When is that not the case?

(My understanding is that `cls` is guaranteed to be either `[]` or `[rep]` and any other value would break Shostak invariants; is that correct?)

Halbaroth replied:
I agree. I reproduced the old code here but we don't need to filter the class at all. This code seems to work:
```ocaml 
    let mk = try ME.find t env.make with Not_found -> assert false in
    let rep, _ = try MapX.find mk env.repr with Not_found -> assert false in
    match X.to_const_term rep with 
    | Some v -> v, ME.add t v mrepr
    | None -> assert false
```
If `X.to_const_term rep` fails to produce a model term here, it's a bug.

bclement-ocp commented on src/lib/structures/expr.ml:
```diff
@@ -3106,3 +3124,20 @@ module BV = struct
+    let rty =
+      match type_info a with
+      | Tfarray (_, rty) -> rty
+      | _ -> invalid_arg "[Expr.ArraysEx.select]"
```
```suggestion
      | _ -> invalid_arg "Expr.ArraysEx.select"
```

bclement-ocp commented on src/lib/util/util.ml:
```diff
@@ -213,3 +212,25 @@ let rec print_list_pp ~sep ~pp fmt = function
   Format.kasprintf (fun s -> raise (Internal_error s)) msg
+
+(* TODO: we have to update the way we manage quoted symbols after
+   the merge of https://github.com/Gbury/dolmen/pull/198 *)
```
This has been merged, I believe (#955 )

bclement-ocp commented on tests/models/arith/arith7.optimize.expected:
```diff
@@ -1,5 +1,7 @@
 
 unknown
+; This model is a best-effort. It includes symbols
+; for which model generation is known to be incomplete.
```
Which ones? This problem only uses the builtin `<` and `<=` on reals, as far as I can tell.

Halbaroth replied:
I didn't fix this issue because I don't understand your point as I explained here: https://github.com/OCamlPro/alt-ergo/pull/925#discussion_r1418918197

bclement-ocp replied:
I replied there (sorry, it got lost int the notifications somewhere) but I don't think this is related; `arith7` doesn't look like it is using any suspicious symbols.

bclement-ocp commented on src/lib/structures/expr.ml:
```diff
@@ -1015,14 +1023,24 @@ let mk_ite cond th el =
 
-let [@inline always] const_term e =
-  (* we use this function because depth is currently not correct to
+let rec is_const_term e =
```
I thought this was replacing the pre-existing `const_term` function. Given that it's not, it is very confusing to have both `const_term` and `is_const_term` performing checks but returning different values. Can we maybe call this `is_model_term` or `is_value_term` instead?

Halbaroth replied:
Sure

bclement-ocp reviewed (commented):
It looks like there are still some issues with the ways arrays and abstract types are printed.

See for instance:

```smt-lib
(set-option :produce-models true)
(set-logic ALL)
(declare-sort S 0)
(declare-const s S)
(declare-datatype Pair ((pair (first (Array Int S)) (second (Array Int S)))))
(declare-const x Pair)
(assert (= (select (first x) 0) s))
(assert (= (first x) (second x)))
(check-sat)
(get-model)
```

This prints:

```smt-lib
unknown
(
  (define-fun s () S (as @a0 S))
  (define-fun x () Pair (pair .k4 .k4))
  (define-fun .k4 () (Array Int S) (store (as @a1 (Array Int S)) 0 s))
)
```

but the `.k4` symbol should not appear in the model as it is an internal name (and only symbols defined by the user should appear in models). The expected output should be along the lines of:

```smt-lib
unknown
(
  (define-fun s () S (as @a0 S))
  (define-fun x () Pair
    (pair
      (store (as @a1 (Array Int S)) 0 s)
      (store (as @a1 (Array Int S)) 0 s)))
)
```

We probably need one last substitution step of fresh array and abstract constants.

Halbaroth committed:
rename `terms_values` to `term_values`

Halbaroth committed:
rebase artefact in models.ml

Halbaroth committed:
Use Dolmen to quote identifiers

Halbaroth committed:
Rename `is_const_term` to `is_model_term`

Variables aren't adequate model values. Now the predicate
`is_model_term` returns `false` on them.

Halbaroth committed:
Remove brackets in `ArrayEx.select`

Halbaroth committed:
Clarify `X.is_constant` implementation

Halbaroth committed:
Simplify `model_repr_of_term`

We don't need to inspect the class of a term to retrieve a constant
expression in it. When we call `model_repr_of_term` during the model
generation, we are sure that the representant of the class is a constant
semantic value and `X.to_const_term` cannot fail. Otherwise, it's a bug!

Halbaroth committed:
Rename `to_const_term` to `to_model_term`

As we rename `Expr.is_const_term` to `Expr.is_model_term`, I rename
this function for consistency purposes.

Halbaroth committed:
Only use `Expr.t` to store model values

As arrays can occur into model values, we need to use `Expr.t` to
represent them.

We cannot produce the appropriate model term with `Arrays.to_model_term`
because we haven't semantic values for arrays. Instead we perform two
passes on a pre-model generating with `X.to_model_term`.
- The first pass collects all the values of arrays and generates a model
  term for each array. If the array was declared by the user, we add
  it to the model.
- The second pass substitutes all the array identifiers in the pre-model
  by model terms we have generated in the first pass.

Halbaroth committed:
Remove warning message for models in presence of FPA symbols

bclement-ocp commented on src/lib/reasoners/shostak.ml:
```diff
@@ -646,8 +676,8 @@ struct
       | Term t, ty      -> (* case disable_adts() handled here *)
-        if Expr.const_term t ||
-           List.exists (fun (t,_) -> Expr.const_term t) eq then None
+        if Expr.is_model_term t ||
+           List.exists (fun (t,_) -> Expr.is_model_term t) eq then None
```
This changes from the old `Expr.const_term` to the new `Expr.is_model_term` — is that intended?

Halbaroth replied:
It is. The function `Expr.const_term` isn't exposed anymore and it's not appropriate for model generation.

bclement-ocp commented on src/lib/reasoners/uf.ml:
```diff
@@ -1099,15 +1035,193 @@ module MED = Map.Make
+  in fun env t ((mdl, mrepr) as acc) ->
+    let { E.f; xs; ty; _ } = E.term_view t in
+    if X.is_solvable_theory_symbol f ty
+    (* TODO: BVand, BVor and BVxor are not symbols of the Shostak theory
+       of bitvectors as the solve function of this theory cannot support them.
+       Still these symbols will appear in the union-find.
+       We should use a different predicate to discriminate appropriate terms
+       here. *)
+    || Sy.equal f Sy.(Op BVand) || Sy.equal f Sy.(Op BVor)
+    || Sy.equal f Sy.(Op BVxor)
```
This can be removed after #1007

bclement-ocp commented on src/lib/reasoners/uf.ml:
```diff
@@ -1099,15 +1035,193 @@ module MED = Map.Make
+              assert false
+          in
+          let mdl =
+            if not @@ Id.Namespace.Internal.is_id (Hstring.view id) then
```
Would checking the `defined` field in the `Sy.Name` make more sense here?

Halbaroth replied:
It doesn't. The `defined` field means the user defined the semantic of the symbols in the input file with `define-fun`. Here we're looking for array identifiers that haven´t been declared by the user. For instance in the input:
```smt2
(set-option :produce-models true)
(set-logic ALL)
(declare-sort S 0)
(declare-const s S)
(declare-datatype Pair ((pair (first (Array Int S)) (second (Array Int S)))))
(declare-const x Pair)
(assert (= (select (first x) 0) s))
(assert (= (first x) (second x)))
(check-sat)
(get-model)
```
We'll produce a identifier for embedded array in `x` but this array isn't declared by the user. Thus we don't want to display the definition of this array at the top level of the model.

bclement-ocp replied:
Right — we'd need `declared`, not `defined`, but we don't have that, of course. Thanks for the explanation

bclement-ocp commented on src/lib/structures/errors.ml:
```diff
@@ -252,6 +257,18 @@ let report_mode_error fmt = function
   | Forbidden_command s ->
     fprintf fmt "Command %s" s
 
+let report_model_error ppf = function
```
I believe raising these indicates an internal error, in which case we should say so in the error message (otherwise the user will not understand why we are trying to substitute these identifiers, since they didn't ask).

Halbaroth replied:
It's not necessary an internal error as the user could apply `subst` on a model while using the library. I prefix the error message with `Model error: ` to clarify where the exception could be raised.

bclement-ocp commented on src/lib/structures/expr.mli:
```diff
@@ -326,8 +335,13 @@ val print_th_elt : Format.formatter -> th_elt -> unit
+    - A record definition involving only model terms.
+    - A constructor application involving only model terms,
+    - A literal of a basic type (integer, real, boolean, unit or bitvector),
+    - An application of function of arity [0]. *)
```
Is this *any* function of arity 0 or a specific subset (I guess internal & abstract IDs? Or maybe just abstract IDs?)

Halbaroth replied:
I thought it could be only abstract and internal names, but it seems to be any name. For instance, 
```smt2 
(set-option :produce-models true)
(set-logic ALL)
(declare-sort s 0)
(declare-const x s)
(declare-fun f (Int) s)
(assert (distinct (f 0) x))
(check-sat)
(get-model)
```
Here, the class representative of `x` is `x` itself. As we use `Expr.is_model_term` in `X.to_model_term`, the name `x` will occur in `Expr.is_model_term`.

bclement-ocp replied:
Ah yes, we don't generate abstract identifiers for things that are already identifiers, there would be no point!

bclement-ocp commented on src/lib/reasoners/arrays.ml:
```diff
@@ -60,28 +60,12 @@ module Shostak (X : ALIEN) = struct
     else
       match X.term_extract r with
       | Some _, true ->
         Some (Expr.fresh_name (X.type_info r), false)
```
Should this be `mk_abstract` now that that exists?

Halbaroth replied:
We could be but it's not very useful. Currently, the only non-regression test calling this part of the code is `embedded-array.models.smt2`:
```smt2
(set-option :produce-models true)
(set-logic ALL)
(declare-sort S 0)
(declare-const s S)
(declare-datatype Pair ((pair (first (Array Int S)) (second (Array Int S)))))
(declare-const x Pair)
(assert (= (select (first x) 0) s))
(assert (= (first x) (second x)))
(check-sat)
(get-model)
```
Applying your suggestion, I tried to retrieve the abstract names returned by `assign_value` but it seems to be complicated. For instance, with the above input file, AE will produces a fresh variables for each field of the pair `x` while generating the model. Let's say we have a fresh name `.k` and the equality `first x = .k` and `second x = .k`. Then we produce our abstract array with `assign_value`, say `@a`. Now, CC(X) uses Shostak's solve function to determine the better class representative and the solve function uses `X.str_cmp` to decide if we have to perform `.k -> @a` or `@a -> .k`. As these names have the same depth, the tag is used to determine the largest one. In this situation `.k` wins the game and becomes the class representative of `first x`.

Of course, we could look in the class of `first x` to retrieve the abstract name `@a` but we don't need to do that most of the time. In general, the class representative is the adequate model value.

In my opinion, the current implementation is fine. The returned value by `assign_value` isn't the definitive value but an internal name used during the model generation.

bclement-ocp replied:
`.k` is a `model_term` (I believe) hence we shouldn't reach this branch that calls `fresh_name` in your example, and so I am not sure I understand your point.

My suggestion to use `mk_abstract` here is that, since we have a concept of abstract identifiers to indicate variables related to model generation, it would make sense to use it for the variables we create during model generation (even if not all of them will end up not in the final model). The current implementation also works, of course :)

bclement-ocp reviewed (commented):
I think we are mostly there; a couple comments mostly on use of internal vs abstract terms & related conditions.

I feel like most identifiers used in model generation should be abstract identifiers (which we will substitute if we learn their actual value) since that's their purpose but there might be issues with this.

Halbaroth committed:
bvor, bvand and bvxor aren't suspicious anymore

Halbaroth committed:
prefix model error

Halbaroth committed:
Poetry

bclement-ocp reviewed (approved):
I'll let you choose whether to use internal identifers or abstract identifiers during model generation; I think on the other points we are ready to merge.
